From a1537804ca633d93da5ffd7fcad00bae25d5c676 Mon Sep 17 00:00:00 2001
From: Gu Zitao <guzitao@wxiat.com>
Date: Tue, 22 Feb 2022 22:12:53 +0800
Subject: [PATCH] sw64: megaraid: fix kernel panic caused by accessing an
 illegal address
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: a1537804ca633d93da5ffd7fcad00bae25d5c676
Modified-by-SEL: No


Sunway inclusion
category: feature
bugzilla: https://gitee.com/openeuler/kernel/issues/I4SPZD
CVE: NA

-------------------------------

On SUBARCH_C3B, when using the original driver for LSI MegaRAID
SAS-3 PCIe card, a kernel panic may arise from accessing an illegal
address. Make some changes for SUBARCH_C3B to avoid it. But it may
not be avoided entirely.

Signed-off-by: Gu Zitao <guzitao@wxiat.com> #openEuler_contributor
Signed-off-by: Laibin Qiu <qiulaibin@huawei.com>
Reviewed-by: Hanjun Guo <guohanjun@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 drivers/scsi/megaraid/megaraid_sas_fusion.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/drivers/scsi/megaraid/megaraid_sas_fusion.c b/drivers/scsi/megaraid/megaraid_sas_fusion.c
index b8ec6db532b0..1cfa2c9db4f2 100644
--- a/drivers/scsi/megaraid/megaraid_sas_fusion.c
+++ b/drivers/scsi/megaraid/megaraid_sas_fusion.c
@@ -3506,6 +3506,14 @@ complete_cmd_fusion(struct megasas_instance *instance, u32 MSIxIndex,
 	       d_val.u.high != cpu_to_le32(UINT_MAX)) {
 
 		smid = le16_to_cpu(reply_desc->SMID);
+#ifdef CONFIG_SUBARCH_C3B
+		if (smid == 0xffff) {
+			smid = d_val.u.low >> 16;
+			if (smid == 0xffff)
+				break;
+		}
+#endif
+
 		cmd_fusion = fusion->cmd_list[smid - 1];
 		scsi_io_req = (struct MPI2_RAID_SCSI_IO_REQUEST *)
 						cmd_fusion->io_request;
-- 
2.26.2

